### PR TITLE
Block Support: Add width block support feature

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related: 
- #28356

Depends on: 
- #32392
- #32499

## Description
- Adds width block support and includes in new Dimensions panel.
- Adds new segmented width control

## How has this been tested?
Style.js test:
`npm run test-unit:watch packages/block-editor/src/hooks/test/style.js`

#### Manual Test Instructions

1. Checkout this PR and build
2. Enable custom dimensions support including height, width, padding and margin - [theme.json gist](https://gist.github.com/aaronrobertshaw/31c37eaca2a17a877ed71e8369e91775)
3. Enable dimensions block support for Group block - [block.json gist](https://gist.github.com/aaronrobertshaw/ea7337b794bc44d7017bf4e82988c756)
4. Create a post, switch to code editor view and paste in the snippet of block code below:
```html
<!-- wp:group {"backgroundColor":"gray"} -->
<div class="wp-block-group has-gray-background-color has-background"><!-- wp:group {"backgroundColor":"white"} -->
<div class="wp-block-group has-white-background-color has-background"></div>
<!-- /wp:group --></div>
<!-- /wp:group -->
```
5. Switch back to the visual editor and select the white inner group block
6. Ensure dimensions panel appears in the sidebar
7. Open the dimension panel's more menu and toggle on the width control
8. Adjust the width and confirm the block's width gets adjusted. Inspect element in dev tools to confirm CSS style
9. Click on the edit button beside the preset percentage widths and configure an explicit width for the group
10. Test clearing the width by toggling off the width control from the panel's ellipsis menu
11. Save post and view on the frontend. The width style should be applied correctly
12. Repeat the process for a dynamic block, e.g. site title, ensuring correct styles on the frontend
13. Update your block.json you've been testing with to set the width block support feature flag to `true` instead of `"segmented"`. Repeat the test process this time ensuring the width control is a standard `UnitControl`

## Screenshots 
WidthSupport-Latest

## Possible Next Steps
- [ ] Possibly add min/max width support alongside this? 

## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by jasmussen** [comment]
This is cool to see, thank you for work on this. Going back to #27331, the design there shows a different width toggle:



I understand that there will be cases where we need precise width control. But arguably, and the reason for that initial design, most of the time you should be using one of the theme-defined widths, such as default, wide, or full-wide. By being default/wide and full-wide, the controls in the sidebar are also aware of what you might have set in the block toolbar, rather than potentially conflict with them (such as choosing full-wide in the block toolbar, and 20px in the inspector).

We almost certainly want the fine-grained width-control for some blocks, such as Button. But probably not group? Or perhaps we need an improved UI that enables both? But in any case, that makes it a bit harder to land this one.

**Comment by aaronrobertshaw** [comment]
> Going back to #27331, the design there shows a different width toggle:

I'll update this. I got a bit of tunnel vision after adding the initial height block support PR and worrying about how these controls would play within the new block support panel (Dimensions in this case). I've added the "In Progress" label to this for the time being.

> We almost certainly want the fine-grained width-control for some blocks, such as Button. But probably not group? Or perhaps we need an improved UI that enables both? But in any case, that makes it a bit harder to land this one.

The group block was only used to test as it is a simple block. While updating this I'll see what options we might have via the block support flags to perhaps opt in to varying levels of width support. Simply opting in might give the segmented control. Setting the support flag to "custom" or something might then allow for the block to be given a UnitControl for that fine-grained control.

**Comment by aaronrobertshaw** [comment]
@jasmussen I've had a chance to revisit this one today however I am struggling to understand how it is supposed to work based on the animated GIF and your previous comment. I didn't find any real answers in https://github.com/WordPress/gutenberg/issues/27331 or https://github.com/WordPress/gutenberg/issues/28356 either so I must be missing something.

My thoughts around the mock up for the width control include:
- It looks like it has just been mislabelled width instead of alignment ( I realise that other mockups include a dropdown for the alignment support options)
- I understand that a block's alignment can impact on it's width. I'm trying to understand how overlapping the width control with the alignment's helps. Especially the reuse of icons.
- Would both the "wide" and "full" buttons from that segmented control actually just be setting 100% width? The alignment support applies a max or min width I believe.
- What would the first button in the segmented control do? Does it set width to `auto` or 50% like the icon visually indicates?
- I can see a lot of users being confused by what that width control is doing. They'll toggle the buttons on and off, then be frustrated that they can't actually control the width as the presence of a width control would indicate.
- Would it not be better for the width controls to update to a meaningful value if the alignment changes to a selection that does impact the block's width? For example, if wide or full alignment is selected then the width is set to 100% or cleared and disabled/hidden.

In short, my current thoughts boil down to:
1. What is the segmented control actually to achieve, what width values can be set via it?
2. How can we reduce the confusion between it and the alignment support?
3. Can we proceed with the initial width support proposed here for explicit widths? It can be used for those blocks that need explicit control and not opted into for blocks that we wish to provide some refined segmented control for?

Thanks in advance for the help getting up to speed on this one 🙏

**Comment by jasmussen** [comment]
> It looks like it has just been mislabelled width instead of alignment ( I realise that other mockups include a dropdown for the alignment support options)

It's a bit forward looking and needs a few other pieces in place, so definitely consider it a mockup more than a fully formed thought. In fact there are a few mockups that go in different directions for exactly the reasons you outline.

But to respond specifically around the alignments/widths mockup, that was intentional enough in that it acknowledges **explicit horizontal pixel widths** as something you should rarely if ever need for most blocks, as they very easily break responsiveness. The alternative are "width toggles" like wide and fullwide, as shown as a width "alignment".

There's another mockup that embraces it as an "alignment", but makes it a dropdown instead, like so:

Frame 251

But even that still needs a bit of thought, with some conversations suggesting we separate the actual alignments from the width based ones, and call those "Positions" instead:

Frame 252

More high level, those designs assume a dimensions panel that is contextual to some of the larger container blocks, like Cover or Group being selected. For other blocks, such as Button or Search, those "positions"/alignments would not make sense, and we'd likely want to end up with something closer to what we have today, with a [segmented control](https://github.com/WordPress/gutenberg/pull/31937) featuring shortcuts to 25%/50%/75%/100% sizes. The balance in the following mockup isn't right — the pixels need massaging, maybe we want to show primarily the segmented control but make explicit values available in a toggle similar to individual paddings? But to convey the point, it's this general thought:

Sidebar

So my previous comment was more to convey that the Dimensions panel needs to be a bit more contextually aware, offering some affordances only to blocks like Search or Button, and alignment affordances to others. Since this PR targets the group, that's why I brought it up. But given we'll almost certainly still need the width control (even if not the height control) for the Button (singular) block, perhaps that's the best place to start? 

If it is specifically destined for the group, we'll need to advance the designs a bit, and today is a bit tricky for me on that front, but I can probably find some time next week.

**Comment by aaronrobertshaw** [comment]
Thank you for the explanations, it's becoming a little clearer 🙇 

> It's a bit forward looking and needs a few other pieces in place, so definitely consider it a mockup more than a fully formed thought.

It sounds like we need to iterate on the designs first before we can go much further down the path of a segmented control.

> But to respond specifically around the alignments/widths mockup, that was intentional enough in that it acknowledges explicit horizontal pixel widths as something you should rarely if ever need for most blocks

Would explicit percentage widths not be perhaps desirable in some cases?

> The alternative are "width toggles" like wide and fullwide, as shown as a width "alignment".

If these toggles overlap with the align `wide` and `fullwide`, they are the same effective width ( 100% ) while the alignment styles set a max width based on the layout. So this still confuses me. Why offer them if they do the same thing and are essentially what the default block width would be without any "width support"?

> There's another mockup that embraces it as an "alignment", but makes it a dropdown instead, like so:

Ok, I interpreted this mockup as only including the separate alignment support, not width. I didn't realise they were supposed to be one and the same in this mockup. 

> But even that still needs a bit of thought, with some conversations suggesting we separate the actual alignments from the width based ones, and call those "Positions" instead:

I'd absolutely vote for separating them.

> maybe we want to show primarily the segmented control but make explicit values available in a toggle similar to individual paddings?

This is something concrete that I could update the width control with in this PR. Do you think it is worth the time doing that now or should I hold off for any further design iterations/decisions?

> Since this PR targets the group, that's why I brought it up. But given we'll almost certainly still need the width control (even if not the height control) for the Button (singular) block, perhaps that's the best place to start?

This PR doesn't actually target the Group block. I simply used it as a block to test with as it's generally a very simple block.

This width control will only be included within the Dimensions panel for blocks that opt into the width block support. If it doesn't make sense for the Group block, we don't opt in for it. If it does make sense for a Button, great, we opt in and so on. Same goes for the height support etc.

If it helps I can rewrite the testing instructions and update the demo GIF to revolve around the individual button block?

**Comment by jasmussen** [comment]
> It sounds like we need to iterate on the designs first before we can go much further down the path of a segmented control.

Cool, I'll find some time next week.

> If these toggles overlap with the align wide and fullwide, they are the same effective width ( 100% ) while the alignment styles set a max width based on the layout. So this still confuses me. Why offer them if they do the same thing and are essentially what the default block width would be without any "width support"?

These toggles in the inspector would be the same as those in the toolbar. But in the mockups they are included because they are conceptually related to _dimensions_. Omitting them from dimensions because they are in the block toolbar already might be confusing. 

In this case, I think of the block toolbar as a shortcut. You can see how that works with the Cover block "full height" option, which when toggled simply sets a 100vh height:

cover

You can see the connection between the two. That's also the key bit about potentially including such alignments in the inspector: wide and fullwide likely need to unset any explicit widths you might have set, because they take precedence. When you see the width becoming unset as you toggle one of those alignments (perhaps the entire widht panel graying out) it might make for a more obvious user experience.

This is part of what needs to be made clearer in the mockups, bug hopefully this comment can set the stage. 

> Would explicit percentage widths not be perhaps desirable in some cases?

Yes, Button does that and it's working well, so we'd at least want it as a default for those. 

We may also want it for the bigger blocks, but we need to find a balance of prominence in the sidebar that naturally leads towards _good choices_ by the user. It's very easy for someone to set a specific width of their block, looking at a desktop interface, and not realizing that width just broke the responsiveness of their site. 

Perhaps that balance is showing alignments by default (whatever we call them), and hiding explicit width in the progressive menu. That would at least enable it for patterns and power users. 

> This is something concrete that I could update the width control with in this PR. Do you think it is worth the time doing that now or should I hold off for any further design iterations/decisions?

Let me see if I can take a quick stab at an overall doodle. 

> This width control will only be included within the Dimensions panel for blocks that opt into the width block support. If it doesn't make sense for the Group block, we don't opt in for it. If it does make sense for a Button, great, we opt in and so on. Same goes for the height support etc.

👍 👍

**Comment by jasmussen** [comment]
Here's a thought:

Padding and width

Inspired by the padding control, there could a _default_ and an _advanced_ view for the width control. By default, it's just shortcuts to useful values such as 25+100%. Click the edit button and the value becomes directly editable.

I'd appreciate a quick sanity check, perhaps @critterverse @javierarce @kjellr? It would be primarily an upgrade to this interface for the Button block:

Screenshot 2021-06-18 at 09 29 10

— but perhaps it could also be a more general Width control you could toggle on for the larger blocks, as discussed in this thread.

Edit: To clarify, this would not be a replacement for all width controls, but be in addition to, something like a "Segmented width control". For something like an Image, we'd still almost certainly need the basic control:

Image

**Comment by kjellr** [comment]
> Screen Shot 2021-06-18 at 11 03 28 AM

I like this. I'm not sure it's a totally clear parallel to the padding control though. For the padding control, you're splitting a value into four sub-values, whereas in the font-size and width controls, you're overriding the preset options. Your mockup feels more akin to the font size control:

Screen Shot 2021-06-18 at 10 58 27 AM

... but more nicely done. 🙂

(As an aside, in the padding mockup from https://github.com/WordPress/gutenberg/issues/27331 I'm not sure folks will know to click on the padding icon, but your width mockup doesn't have that issue.)

**Comment by aaronrobertshaw** [comment]
The proposed segmented width control looks great! ✨ 

I'll work on a prototype for it in this PR along with a means to choose which type of width control make sense when opting into the width block support. 

> Edit: To clarify, this would not be a replacement for all width controls, but be in addition to, something like a "Segmented width control". For something like an Image, we'd still almost certainly need the basic control:

My current thinking is that the feature flag for the block support would accept a couple of different values rather than only a boolean. If it is set to `true` or `"basic"` then it will display a standard UnitControl. If it is set to `"segmented"` then it will display this new control.

> These toggles in the inspector would be the same as those in the toolbar.

Ok, so this means that the block toolbar controls are alignment block support controls and so to are the controls in the inspector.

This is where I think I've failed to communicate the confusion I've had. Previous descriptions and the mockup labelled alignment controls as width controls leading me to believe we needed to be overlapping their functionality. 

My question about omitting options from the width block support controls was to remove some of that overlap thinking that the alignment block support controls would cover those elsewhere in the dimensions panel. Whether that was in a segmented control or the proposed dropdown.

> You can see the connection between the two. 

If the alignment block support adds controls to the dimensions panel as mentioned above, the inspector controls there would inherently match or be "connected" to the alignment options in the block toolbar as they'd both use the same block attribute.

---

My current understanding that I'll move forward with is:

- Alignment and width block support are separate and different
- The one exception to the above is that the width block support value would be cleared if align `wide` or `full` gets set
- The controls for alignment and width block supports are separate
- This is the simplest, cleanest, and, easiest approach to move forward with that can be iterated on if desired

If the above is correct, then all that is required on this PR is to: 

- add the new proposed width control (the one with preset % widths and toggle to switch to UnitControl + slider)
- allow selection of the type of width control to be displayed when opting into block support

**Comment by jasmussen** [comment]
My reason for thinking this approach of two width controls can work is that:

- For Button blocks and others, we'll still want those uniform percentage width shortcuts, as they don't break responsiveness
- It would be nice to also offer advanced control of the width, but with intentionally reduced prominence 
- We will still want explicit width/height controls for Images, side by side as is a custom, for blocks like Image

That's the primary problem we want to solve. 

The thing about the design I'd like to still noodle on Kjell brings up. I'd like a stronger pattern for this type of _toggling between basic and advanced_, I'm not sure the edit button is right. I'm not sure we want a slider for the width. I'd love to find a pattern that is more easily scalable to other controls that might swap between basic and advanced like this. So I think we probably want this, but I'd also love a small sanity check by @mtias if he has time.

**Comment by aaronrobertshaw** [comment]
I've created a prototype for the proposed width control as part of this PR. 

The width block support now allows for opting into this new control in place of a standard `UnitControl` via the block.json file. This means we could use a standard width control for an image and the new control for the button etc.

> I'm not sure we want a slider for the width

I think you are right here, the slider makes sense when we have more concrete boundaries like `0` and `100` for percentage widths. Things get more complicated when using non-percentage units, blocks having custom layouts of varying widths, be nested to various levels etc.

The prototype has a `RangeControl` slider at the moment. It can easily be removed if no longer desired.

| With Slider | Without Slider |
|---|---|
| WidthControl | WidthControl2 |

**Comment by aaronrobertshaw** [comment]
In case it aids further reasoning about the width control and width support I have a separate PR trialling adding the width block support and new control to the button block. https://github.com/WordPress/gutenberg/pull/32878

**Comment by jasmussen** [comment]
Thanks for your patience. I'm getting some notices running this branch, probably a result of my local env (`Notice: register_block_type_from_metadata() was called incorrectly. Block "core/paragraph" is declaring lineHeight support in block.json file under supports.lineHeight`), so I can only look at your GIFs.

And from what I can see, I think it's worth trying:

- We need a custom width panel for the image.
- We need a "shortcut" width panel for Button.
- It'd be nice to be able to switch from the latter to the former.

As for next steps, outside of a good code review:

- Is there any chance you could use the segmented control from #31937?
- Agreed, let's try without the slider.
 
My biggest gotcha with this one is that when you click the edit icon to "unlock", you have to "tab backwards" to set focus in the width panel.

**Comment by aaronrobertshaw** [comment]
>  I'm getting some notices running this branch, probably a result of my local env

I've started fresh, checking out this branch, building it, then following through the test setup as per PR description (steps 1-3). I can't replicate the issues you were facing. Opting into the width support for the paragraph block specifically, since that was mentioned in the error, worked fine as well (not suggesting we want support on that though).

My guess is if you were switching between branches with the build watch task running something went awry. Rebuilding this branch should sort it out.

> We need a custom width panel for the image.

I'm not sure I follow what you are looking for here. Do you mean we want the simple UnitControl display of the width control for the image block? Or, that we need a completely custom dimensions panel for the image block? One that allows for the size control etc.

> We need a "shortcut" width panel for Button.

By "shortcut" do you mean the version of the width control with the preset percentage widths? The panel for me is the dimensions panel. This PR is about the width block support and the control to support that.

> It'd be nice to be able to switch from the latter to the former.

I take it that you are not referring to the toggle button to switch from the preset button group display of the width control to the basic unit control display. 

At present, this width block support allows you to choose how the width control is presented. There are two options:

| Basic UnitControl | New Segmented Control |
|---|---|
|  <p>Block.json opts in with `width: true`</p> Screen Shot 2021-06-23 at 11 47 59 am |  <p>Block.json opts in with `width: "segmented"`</p>WidthControl3 |

> Is there any chance you could use the segmented control from #31937?

Once that PR is finished and lands, we can definitely update this width control. It can be done separately to this PR. I imagine we'll want to also update other uses of button groups that can be replaced by that segmented control.

> Agreed, let's try without the slider.
> My biggest gotcha with this one is that when you click the edit icon to "unlock", you have to "tab backwards" to set focus in the width panel.

I'll remove the slider from the new width control. It should also be possible to set focus to the unit control after "unlocking" the basic view via the edit icon.

**Comment by aaronrobertshaw** [comment]
The `RangeControl` has been removed from the "custom" (explicit width) view in https://github.com/WordPress/gutenberg/pull/32502/commits/d62467fb89f92b718c798ec7c7d9be0f9305dacd

When switching to the custom view now, focus will be moved to the `UnitControl`. https://github.com/WordPress/gutenberg/pull/32502/commits/5ab0a7070e13d924deb42535bd95d922a29eaaff

**Comment by jasmussen** [comment]
> I'm not sure I follow what you are looking for here. Do you mean we want the simple UnitControl display of the width control for the image block? Or, that we need a completely custom dimensions panel for the image block? One that allows for the size control etc.

I was unclear here. I was simply confirming your thoughts: we will need the shortcuts from the segmented control with some blocks, and we'll need just a basic width input field for others. 

Your two options look good to me to try out 👍 👍

**Comment by paaljoachim** [comment]
What are the next steps in this PR?
Thanks!

**Comment by aaronrobertshaw** [comment]
> What are the next steps in this PR?

Thanks for checking in @paaljoachim 👍 

The [height](https://github.com/WordPress/gutenberg/pull/32499) and width block supports weren't in the must-haves for 5.9 so were put on the back burner until recently. 

Unfortunately, with the 5.9 release, we didn't have a means of overriding the theme.json classes to allow the addition of new block support features. This is currently being addressed with the required updates being backported to 5.9.1.  As you would expect this has been a blocker to progressing this PR.

That said, I've been working on bringing the new block support PRs up-to-date and should be in a position to push those changes sometime this week.

In summary, the next steps are:

- Finish debugging remaining issues/requirements for extending theme.json classes
- Update the height support PR
- Update this width support PR ( it currently is based on the height PR )
- Extract the width control into its own component within the components package

**Comment by paaljoachim** [comment]
Thanks for the update @aaronrobertshaw 
It is much appreciated.

**Comment by aaronrobertshaw** [comment]
> That said, I've been working on bringing the new block support PRs up-to-date and should be in a position to push those changes sometime this week.

There are ongoing issues with being able to extend the theme.json settings which are preventing updating this PR. It may not be completed this week. I'll also be AFK until mid-March so there might be further delays on progressing width support.

**Comment by paaljoachim** [comment]
It sounds like solving a few of the issues with the theme.json before landing this PR would be helpful.
Have a nice break Aaron!

**Comment by aaronrobertshaw** [comment]
I've cherry-picked some commits from https://github.com/WordPress/gutenberg/pull/38883 to allow the proper extension of theme.json and global styles. 

At this time, this PR is working however there has been a recent change of plans around how we might manage the general layout of blocks. This might actually include new layout blocks specifically. It is yet to be seen how appropriate control over height and width will be within this context. 

As a result, I will close this PR and the other related dimensions support PRs.(https://github.com/WordPress/gutenberg/pull/32499, https://github.com/WordPress/gutenberg/pull/32878)

**Comment by paaljoachim** [comment]
Regarding...

> At this time, this PR is working however there has been a recent change of plans around how we might manage the general layout of blocks. This might actually include new layout blocks specifically. It is yet to be seen how appropriate control over height and width will be within this context.

Is it possible to link to issues/PR's/more information about these things?
Thanks!

**Comment by aaronrobertshaw** [comment]
Thanks for the question @paaljoachim. At this stage, I don't have any specific issues, PRs, or additional information about the layout blocks. Unfortunately, it is a bit too early for that and we'll need to do some exploration and experimentation first.

About the best I can do is offer this [link](https://every-layout.dev/layouts/) highlighting the various layouts we'll be looking to offer whether by block supports/theme.json or dedicated layout blocks.

**Comment by paaljoachim** [comment]
Interesting! Thank you @aaronrobertshaw
